doc/source/development/rfc/rfc91_dataset_close.rst

```diff
@@ -0,0 +1,184 @@
+RFC 91: GDALDataset::Close() method
+=============================================================
+
+============== =============================================
+Author:        Even Rouault
+Contact:       even.rouault at spatialys.com
+Started:       2023-Jan-20
+Status:        Draft
+Target:        GDAL 3.7
+============== =============================================
+
+Summary
+-------
+
+This RFC aims at providing more robust error detection during dataset closing,
+in particular for datasets in creation/update mode.
+
+It modifies the GDALDataset::FlushCache() method and GDALFlushCache()
+function to return an error code, adds a GDALDataset::Close() virtual method and
+modifies GDALClose() to return an error code.
+
+Motivation
+----------
+
+A lot of drivers that have write capabilities may do write operations in their
+destructors, or just close the file descriptor owned by the dataset. Currently
+there is no clean way of catching errors that might occur. This recently bite
+`Fiona <https://github.com/Toblerity/Fiona/issues/1169>` where the GeoJSON driver
+may emit an error in its dataset destructor when writing the content of a file
+to a cloud object storage.
+
+Code that currently wants to detect if GDALClose() causes an error has to
+do something like the following:
+
+.. code-block:: c++
+
+    CPLErr eErrCodeBefore = CPLGetLastErrorType();
+    GDALClose(hOutDS);
+    if( eErrCodeBefore == CE_None && CPLGetLastErrorType() != CE_None )
+        printf("Error occurred during GDALClose()!\n")
+
+
+This is not pretty and there is always the slight risk that something in the
+chain of calls would clean the error state. It is thus desirable for the
+GDALClose() method to return an error code.
+
+Technical details
+-----------------
+
+The GDALDataset::FlushCache() method is modified to return a CPLErr return value.
+
+A new virtual method GDALDataset::Close() is added.
+
+.. code-block:: c++
+
+    /** Do final cleanup before a dataset is destroyed.
+     *
+     * This method is typically called by GDALClose() or the destructor of a
+     * GDALDataset subclass. It might also be called by C++ users before
+     * destroying a dataset. It should not be called on a shared dataset whose
+     * reference count is greater than one.
+     *
+     * It gives a last chance to the closing processus to return an error code if
+     * something goes wrong, in particular in creation / update scenarios where
+     * file write or network communication might occur when finalizing the dataset.
+     *
+     * Implementations should be robust to this method to be called several times
+     * (on subsequent calls, it should do nothing and return CE_None).
+     * Once it has been called, no other method than Close() or the dataset
+     * destructor should be called. RasterBand or OGRLayer owned by the dataset
+     * should be assumed as no longer being valid.
+     *
+     * If a driver implements this method, it should also call it from its
```

Comment:
Don't give wiggle room: it should also call -> it must also call

Reply:
done